### PR TITLE
tests: port integration tests to spread

### PR DESCRIPTION
**Pull request by mvo5**
This ports the integration tests to spread. Unfortunately the writablepaths test fails with the current image and the grub test also fails. Those are real errors that we need to fix in the image.

I have not (yet) ported the ubuntufan test, that looks like more work, especially since it is currently using docker. The update-rollback-stresstest is also missing, that will get its own branch.

### CODE REVIEW

File: tests/main/ubuntu-core-snap-persists/task.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+summary: Ensure a snap is still there after reboot
```
Comment (niemeyer) [inline tests/main/ubuntu-core-snap-persists/task.yaml]: That's a bit of a strange test. Feels too expensive for what it is testing.

Can we merge this check into something else that needs to reboot for better reasons?  The service test above, maybe?

File: tests/main/ubuntu-core-service/task.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+environment:
+    # FIXME: use test-snapd-tools once it has a service
+    SNAP: xkcd-webserver
+    SERVICE: xkcd-webserver
+execute: |
+    assert_service_running() {
+        systemctl status snap.$SNAP.$SERVICE.service|grep 'Active: active (running)'
+    }
+    if [ "$SPREAD_REBOOT" = "0" ]; then
+        snap install $SNAP
```
Comment (niemeyer) [inline tests/main/ubuntu-core-service/task.yaml]: Moving this to prepare, we can do:
```
prepare: |
    snap install $SNAP
execute: |
    systemctl status snap.$SNAP.$SERVICE.service|grep 'Active: active (running)'
    if [ $SPREAD_REBOOT = 0 ]; then
        REBOOT
    fi
```

File: tests/main/ubuntu-core-service/task.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+summary: Ensure that service keep working after a reboot
+systems: [ubuntu-core-16-64]
+environment:
+    # FIXME: use test-snapd-tools once it has a service
```
Comment (niemeyer) [inline tests/main/ubuntu-core-service/task.yaml]: We have tests/lib/network-bind-consumer with a simple daemon.

File: tests/lib/prepare.sh [hunk 4]
```diff
@@ -50,6 +50,11 @@ setup_reflash_magic() {
         echo "$want_pw" > /tmp/new-shadow
         tail -n +2 /etc/shadow >> /tmp/new-shadow
         cp -v /tmp/new-shadow $UNPACKD/etc/shadow
+
+        # ensure spread -reuse works in the core image as well
+        if [ -e /.spread.yaml ]; then
+            cp -av /.spread.yaml $UNPACKD
```
Comment (niemeyer) [inline tests/lib/prepare.sh]: Hah, good point.

File: tests/main/ubuntu-core-writablepaths/task.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    echo "Ensure everything in writable-paths is actually writable"
+    missing=""
+    broken=""
+    while read -r line; do
+        line=$(echo $line | sed -e '/\s*#.*$/d')
+        if [ -z "$line" ]; then
+            continue;
+        fi
+        # a writable-path may be either a file or a directory
+        #
```
Comment (niemeyer) [inline tests/main/ubuntu-core-writablepaths/task.yaml]: Del line?

File: tests/main/ubuntu-core-writablepaths/task.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            missing="$dir_or_file\n$missing"
+        elif [ -f "$dir_or_file" ]; then
+            if ! touch "$dir_or_file"; then
+                broken="$dir_or_file\n$broken"
+            fi
+        elif ! touch "$dir_or_file"/random-name-that-I-made-up; then
+            broken="$dir_or_file\n$broken"
+        fi
+        rm -f $dir_or_file/random-name-that-I-made-up
+    done < /etc/system-image/writable-paths
```
Comment (niemeyer) [inline tests/main/ubuntu-core-writablepaths/task.yaml]: Can we move this ahead, with `cat /etc/system-image/writable-paths | \`? Having at the end means not quite understanding what we're parsing until we find out the stdin coming.

**Comment by niemeyer** [comment]
A few notes and LGTM assuming tests pass (currently broken).

File: tests/main/ubuntu-core-snap-persists/task.yaml (hunk 1 above)
Comment (mvo5) [inline tests/main/ubuntu-core-snap-persists/task.yaml]: Sure, done

**Comment by niemeyer** [comment]
Merged the #1761 follow-up which had green tests.

**Comment by niemeyer** [comment]
Uh oh, I lied. The #1761 follow up, although based on this, misses 6aaa1b5.

Re-running tests.

**Comment by mvo5** [comment]
Hrm, hrm, this works for me!

**Comment by mvo5** [comment]
I will try this again and if it hangs again i need help from someone who can look at a linode console.

**Comment by niemeyer** [comment]
@mvo5 I've investigated a bit and found the root cause for the problems. We're reaching a limit of reboots in Linode's automatic reboot system:

> To prevent a loop, Lassie will give up if there have been more than 5 boot jobs issued within 15 minutes.

Which I guess is reasonable for normal systems. I'll need to tweak spread is it revives the machine.

**Pull request merged**